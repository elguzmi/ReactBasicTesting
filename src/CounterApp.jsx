// rafc   = crear un functional component rapidamente
import { useState } from 'react'  // use significa hook , es una funcion especial
import PropTypes from 'prop-types'
 

// Los hooks son funciones
// eslint-disable-next-line react/prop-types
const CounterApp =({ counter })=>{  // la property no se cambia como tal
    const [ count  , setCount ] = useState( counter ); // desestructuracion   -> 0 es el valor del estado
    // fucniones flecha son mas rapidas y no cambian el valor del this
    const handleAdd = (event , message)=> {
        // cambio en el componenten y tiene que redibujar esos cambios, toca decirle a react que redibuje el functional component
        //setCount( count + 1); // decirle a react que el estado cambio y va a renderizar el componente
        setCount((c)=> c+ 1)
    }
    const handleSustract = ()=> setCount( count -1 );
    const handleReset = ()=> setCount(counter);
    return (
        <>
            <h1>Counter App</h1>
            <p>{ count }</p>
            <button  aria-label='btnSuma' onClick={ (event)=> handleAdd(event, 'Hola') }> +1 </button>
            <button onClick={ handleSustract }> -1 </button>
            <button aria-label='btn-reset' onClick={ handleReset }> reset </button>
        </>
    )
}


CounterApp.proptypes ={
    counter : PropTypes.number.isRequired
}
CounterApp.defaultProps={
    counter : 0
}

export default CounterApp