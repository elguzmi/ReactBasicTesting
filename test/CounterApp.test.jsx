import { fireEvent, render , screen  } from "@testing-library/react"  // screen el objeto que estamos renderizando
import CounterApp from "../src/CounterApp"

/* eslint-disable no-undef */
describe( 'Pruebas en <CounterApp>', ()=>{    
    const initialNumber = 100;
    test('debe tomar un snaptshot y compararlo', () => {
        const { container } =  render( <CounterApp counter={initialNumber}></CounterApp>)
        expect( container).toMatchSnapshot();
    })

    test('Debe de mostrar el valor inicial de 100', () => {
        render( <CounterApp counter={initialNumber}></CounterApp>)
        //expect(parseInt(screen.getByText(initialNumber).innerHTML)).toBe(100)
        expect( screen.getByText(100)).toBeTruthy();
    })
    test('debe de incrementar con el boton +1', () => {
        render( <CounterApp counter={ initialNumber }></CounterApp>)
        fireEvent.click( screen.getByText('+1') )  // disparar un evento
        //fireEvent.click( screen.getByText('+1') )  // disparar un evento
        expect( screen.getByText('101')).toBeTruthy();
        
    })
    test('debe de decrementar con el boton -1', () => {
        render( <CounterApp counter={ initialNumber }></CounterApp>)
        fireEvent.click( screen.getByText('-1') )  // disparar un evento
        //fireEvent.click( screen.getByText('+1') )  // disparar un evento
        // screen.debug()  // el estado del dom lo podemos ver en consola
        expect( screen.getByText('99')).toBeTruthy();
    })
    test('Debe de funcionar el boton de reset', () => {
        render( <CounterApp counter={ initialNumber }></CounterApp>)
        fireEvent.click( screen.getByText('+1') )  // disparar un evento
        fireEvent.click( screen.getByText('+1') )  // disparar un evento
        fireEvent.click( screen.getByRole('button', {name: 'btn-reset'}) )  // disparar un evento
        // por name (area-label ) o por Id
        screen.debug()
        expect( screen.getByText( '100' ) ).toBeTruthy()
        // 103
    })
    
    
    
    
})