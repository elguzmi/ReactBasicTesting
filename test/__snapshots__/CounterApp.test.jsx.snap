// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <CounterApp> debe tomar un snaptshot y compararlo 1`] = `
<div>
  <h1>
    Counter App
  </h1>
  <p>
    100
  </p>
  <button
    aria-label="btnSuma"
  >
     +1 
  </button>
  <button>
     -1 
  </button>
  <button
    aria-label="btn-reset"
  >
     reset 
  </button>
</div>
`;
